syntax = "proto3";

package stock;

// Message used to store the user data
// in the statefun and also what we give back to the
// queries
message ItemData {
    uint64 id = 1;
    int64 price = 2;
    int64 amount = 3;
}

// Message used to keep track of the current max
// id to be assigned
message Count{
    uint64 num = 1;
}


// Create an item with a certain price.
message CreateItemRequest{
    int64 price = 1;
    string request_id = 2;
    string worker_id = 3;
}

// internal to be used by the statefun
// between the create_item and the item function
message CreateItemWithId{
    uint64 id = 1;
    int64 price = 2;
    string request_id = 3;
    string worker_id = 4;
}

message StockRequest {

    // Get stock item.
    message FindItemRequest{
        uint64 id = 1;
    }

    // Substract stock.
    message SubtractItemStockRequest{
        uint64 id = 1;
        int64 amount = 2;
    }

    // Add stock.
    message AddItemStockRequest{
        uint64 id = 1;
        int64 amount = 2;
    }

    // Request id for identifying the request
    // in the endpoint and answer
    string request_id = 1;

    // Worker id used as key of the response
    // so a given worker just has to save its
    // own messages
    string worker_id = 2;

    oneof message {
        FindItemRequest find_item = 3;
        SubtractItemStockRequest substract_stock = 4;
        AddItemStockRequest add_stock = 5;

    }
}

message StockResponse {
    string request_id = 1;
    string item_id = 2;
    string result = 3;
}